sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/BroadcastExchangeExec.scala

```diff
@@ -58,11 +58,21 @@ case class BroadcastExchangeExec(
     BroadcastExchangeExec(mode.canonicalized, child.canonicalized)
   }
 
+  @transient
+  private lazy val promise = Promise[broadcast.Broadcast[Any]]()
+
+  /**
+   * For registering callbacks on `relationFuture`.
+   * Note that calling this field will not start the execution of broadcast job.
+   */
+  @transient
+  lazy val completionFuture: scala.concurrent.Future[broadcast.Broadcast[Any]] = promise.future
+
   @transient
   private val timeout: Long = SQLConf.get.broadcastTimeout
 
   @transient
-  private lazy val relationFuture: Future[broadcast.Broadcast[Any]] = {
+  private[sql] lazy val relationFuture: Future[broadcast.Broadcast[Any]] = {
```

Comment:
can we just change this to scala future?

Reply:
Scala future is not cancellable :(

Reply:
Yeah, just to include more info in our offline discussion here:

What is needed by AQE is a "cancellable" Future with callbacks.

Java Future is cancellable, but has no callbacks.
Scala Future has callbacks, but is not cancellable.
The Spark impl of a cancellable Scala Future (with callbacks), namely `FutureAction`, is more specific for Spark jobs and for general purpose, thus not so friendly to `BroadcastExchangeExec` here.

Looks like an option can be to implement a general purpose `FutureAction` to replace the Java Future in `BroadcastExchangeExec`. Let's just mark this a follow-up though.